Pathfinding unit tests:

ximinez opened the pull request:
* Refactor ripple path find to be more testable.
* Reimplements the first 4 tests from `tests\path-test.js`
* Verify balances in Ledger test.

Reviewers: @HowardHinnant, @scottschurr, @rec 

By way of background, my intention with these changes was to make debugging my changes for RIPD-655 easier. In the process of developing and validating the test functions, I started reimplementing the tests from `tests/path-test.js`. Even though it's not done, with RIPD-655 on hold for now, I figure the utility functions may be useful for other path-related tests. There are no RIPD-655-related changes in this PR. (Unsquashed changes are in https://github.com/ximinez/rippled/tree/x-path-unit.)

rec commented:
1. Quite a lot of lines are quite a bit longer than 100 characters.
2. We're now naming C++ test files as `.test.cpp`.  I was fine with `_test.cpp` but there we are.

sublimator commented:
I agree with Tom. If the idea was to make it easier for David (the next guy) to write tests to aid in his developing/fixing path finding, I think we could go further to achieve that goal.

There's lot of copy/paste here, eg:
```
Ledger::pointer LCL = createGenesisLedger(100000 * xrp, master);
Ledger::pointer ledger = std::make_shared<Ledger>(false, *LCL);
```

And when I see this block:
```
   expect(alternatives.size() == 1, "result.second.size() == 1");
```

It only reinforces my belief we should use MACROs for the expects.  `result.second.size` is stale and unmaintained. There's no `result` in scope. The "next guy" will get an error like `#12 result.second.size() == 1` when it could link to a file/line annotated expect expression, that doesn't need to be maintained separately to some string. 

The strings could be compiled out in release builds if we really worried about binary bloat, by removing the annotations.

scottschurr commented:
Hi Ed, I've been charging ahead with multi-sign tests while you were working on these.  Of course our paths have diverged, in part because we have different requirements.  An important consideration for me is that transactions not arrive pre-signed, since I need to sign them in a variety of ways.  You don't have that constraint.

At any rate, if you're curious, you can see where I've gone by looking at these three files:
https://github.com/scottschurr/rippled/blob/m-of-n/src/ripple/app/transactors/test/common_transactor.h
https://github.com/scottschurr/rippled/blob/m-of-n/src/ripple/app/transactors/test/common_transactor.cpp
https://github.com/scottschurr/rippled/blob/m-of-n/src/ripple/app/transactors/test/MultiSign.test.cpp
At some point we should have a discussion to see if it makes sense to have one infrastructure that we can share.  But that discussion can wait until after your pull request is merged.

HowardHinnant commented:
This P/R builds with no warnings on OS X. This was built using `scons debug` which builds both unity and non-unity builds. The unity debug build passes unittest, npm, syncs and stops.

 I have no comments beyond what have already been made.  I'm happy to work on some refactoring along the lines of what Tom suggested.
:+1:

scottschurr commented:
Ooooh.  The bottom of this pull request looks like a pretty serious refactoring of path finding.  I can look it over, but I'm in no position to say whether it's correct.  I hope Howard or Tom can add some confidence there.

ximinez commented:
@rec I can't reply nicely inline, so I'll quote:

>    Quite a lot of lines are quite a bit longer than 100 characters.

Cleaned up.

>    We're now naming C++ test files as .test.cpp. I was fine with _test.cpp but there we are.

We are!?!? I thought we were changing from .test.cpp to _test.cpp!

nbougalis commented:
@ximinez: `.test.cpp` and `_test.cpp` are both fine.

ximinez commented:
@sublimator I can't reply nicely inline, so I'll quote:

> There's lot of copy/paste here, eg:

Yes, there is. The example mismatched `expect` wasn't the only one either, so I made a couple of changes:

1. Changed the base class, and rewrote most of the `expect` to use `TestSuite::expectEquals`.
2. Rewrote `createGenesisLedger` to return a pair, and `close_and_advance` to take references and return `void`. I think @rec's suggested test harness is a little too heavy, but this will at least some of the more _pointless_ copypasta.

scottschurr commented:
Looks okay, although I hope you choose to cap your line lengths at 80 characters.  I had some comments, but I didn't spot any clearly wrong code.  I have zero confidence in my review of the path finding refactor.  I hope someone else can comment sensibly on that part.
:+1:

ximinez commented:
@scottschurr 

> The bottom of this pull request looks like a pretty serious refactoring of path finding.

It is not as serious as it looks. My goal was to refactor some code out of `doRipplePathFind` into `buildSrcCurrencies` and `ripplePathFind` *without changing logic* so that tests could call the new functions without needing as many dependencies.

scottschurr commented:
@ximinez wrote "My goal was to refactor some code ... without changing logic"

I guessed that was the case.  I'm just saying that the refactor was serious enough that I was unable to follow how the pieces from the original code moved to their new resting places.  That doesn't make the refactor wrong, it just means I'm not confident in my review of that portion of your changes.

ximinez commented:
@scottschurr Yeah, that's a big deficit of diffs on github. Try from a command line: `git diff -w c3d200..4cc43` or `gitk` with the "Ignore space change" checkbox selected, and see if it doesn't make a little more sense.

scottschurr commented:
@ximinez: Good advice.  I used kdiff3 to look at the path finding differences and it made a lot more sense.  I now have more confidence in the path finding refactor.  And thanks for fixing up the line lengths.
:+1:

ximinez commented:
@scottschurr 

> Of course our paths have diverged, in part because we have different requirements.

Boy, have we! What's more, I think you've gone a long way toward implementing the test harness that @rec suggested above.

> An important consideration for me is that transactions not arrive pre-signed, since I need to sign them in a variety of ways. 

The only mitigation I can offer there is that the various overloads of `getPaymentTx` take a `sign` parameter that allows you to disable signing (though all of the tests in Path_test.cpp use signed transactions, Ledger_test.cpp has test cases that don't). The returned `STTx` can be manipulated further before being given to `applyTransaction`. If the `STTx` is too opaque, there's also a `getPaymentJson` helper.

> At some point we should have a discussion to see if it makes sense to have one infrastructure that we can share.

Definitely. Also include @HowardHinnant, since I think he'll be the most likely to next work directly on these tests. The reason I'm submitting these now is that the relevant ticket is being put on hold for a while, and I don't want these changes to go too stale before I get back to them, so I may not even need to be included unless you have questions or need help.

scottschurr commented:
Okay, there's a plan.  Assuming the m-of-n pull request ever makes it into develop ;-), I'll talk with Howard  about whether it makes sense to have one common test infrastructure for these sorts of business logic tests.  If it does, then Howard and I will negotiate which approach is preferable and go from there.

HowardHinnant commented:
Sounds like a plan.

What's with the travis failures?!

ximinez commented:
> What's with the travis failures?!

I have no idea! I just started seeing the same error from appveyor, too, but everthing is working perfectly locally.

```ripple.ripple_app.Path payment auto path find (using build_path)
2015-Mar-17 22:09:37 SHAMap:WRN SHAMap setItem, no change
2015-Mar-17 22:09:37 STObject:WRN setType( ) invalid default Paths
2015-Mar-17 22:09:37 STTx:WRN Transaction not legal for format
#2 failed: unhandled exception: transaction not valid```

HowardHinnant commented:
I'm building locally on OS X now...

HowardHinnant commented:
I've got an npm failure on OS X.
````
1) Order Book Track offers:
     Uncaught TypeError: Cannot call method 'ifError' of undefined
````
This is new.  I did not have this 3 hours ago on this PR.  I just did a `npm install`, and the error went away.

nbougalis commented:
@HowardHinnant: I'll verify the `package.json` file for this repom but chances are that switching between branches got npm confused with two different versions of `assert-diff`.

ximinez commented:
@nbougalis @HowardHinnant , Yeah, I've seen similar errors when my modules are out of date. Pretty sure that one's spurious.

...Which makes the travis and appveyor failures all the more frustrating, because I can't find a way to reproduce them!

ximinez commented:
The Travis failure must be spurious, because the same code is succeeding in Travis when built directly from my branch. https://travis-ci.org/ximinez/rippled/builds/54811645

Maybe it'll straighten out when I rebranch.

ximinez commented:
Travis mystery solved: The issue arises when my changes are merged with the latest develop. This branch is based off of b14, develop is b15. I'll figure out _why_ tomorrow.

vinniefalco commented:
"We're now naming C++ test files as .test.cpp. I was fine with _test.cpp but there we are."

Nope, we aren't. All that is required is that they reside in a directory called "tests".

ximinez commented:
I found the problem causing the Travis failures. b15 changed the default behavior of the rippling flag, so I need to reenable it for any tests that rely on it. It was fixed in the JS tests at https://github.com/ripple/rippled/compare/c3d200d...ripple:ef2a436#diff-d1b2ee28d3a6a21680f9c09df1602431R362.

Does anybody have opinions on whether I should close this PR, fix it, and open a new one, or leave this PR open (and set the Hold label), fix it, and rebase?

PS. "8"?
